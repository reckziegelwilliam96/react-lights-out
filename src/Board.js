import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/



function Board({ nrows=5, ncols=5, chanceLightStartsOn=0.5 }) {
  const [board, setBoard] = useState(createBoard(chanceLightStartsOn));

  function cellStartsOn(chance) {
    return Math.random() < chance;
  }

  function createBoard(chanceLightStartsOn) {
    let initialBoard = [];
    for (let x = 0; x < nrows; x++) {
      let row = [];
      for (let y = 0; y < ncols; y++) {
        row.push(cellStartsOn(chanceLightStartsOn));
      }
      initialBoard.push(row);
    }
    return initialBoard;
  }

  function hasWon(board) {
    for (const row of board) {
      for (const cell of row) {
        if (cell) {
          return false;
        }
      }
    }
    return true;
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };


      const copy = oldBoard.map((row) => row.slice());

      flipCell(y, x, copy);
      flipCell(y + 1, x, copy);
      flipCell(y - 1, x, copy);
      flipCell(y, x + 1, copy);
      flipCell(y, x - 1, copy);

      return copy;
    });
  }
  const isGameWon = hasWon(board);

  // Render the game board
  const renderBoard = () => {
    return (
        <tbody>
          {board.map((row, y) => (
            <tr key={y}>
              {row.map((cell, x) => (
                <Cell
                  key={`${y}-${x}`}
                  isLit={cell}
                  flipCellsAroundMe={() => flipCellsAround(`${y}-${x}`)}
                />
              ))}
            </tr>
          ))}
        </tbody>
    );
  };

  return (
    <table className="Board">
      {isGameWon ? <h2 className="Board-winning">You won!</h2> : renderBoard()}
    </table>
  );

}

export default Board;
