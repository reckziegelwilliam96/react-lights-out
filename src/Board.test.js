import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Board from './Board';

test('renders Board with a predefined initial state', () => {

  const { container } = render(
    <Board nrows={3} ncols={3} chanceLightStartsOn={0.5} />
  );

  expect(container).toMatchSnapshot();
});

test("flips the right cells when a cell is clicked", () => {
    const { container } = render(<Board nrows={5} ncols={5} chanceLightStartsOn={1} />);
  
    const litCellsBeforeClick = container.querySelectorAll(".Cell-lit");
  
    fireEvent.click(litCellsBeforeClick[12]); // Click on the center cell
  
    const litCellsAfterClick = container.querySelectorAll(".Cell-lit");
  
    expect(litCellsAfterClick.length).toBe(20);
  });