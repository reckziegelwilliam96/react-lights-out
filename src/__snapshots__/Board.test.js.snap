// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Board with a predefined initial state 1`] = `
<div>
  <div
    class="Board"
  >
    <table
      class="Board"
    >
      <tbody>
        <tr>
          <td
            class="Cell "
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell "
          />
          <td
            class="Cell "
          />
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
